core/utils/utils.go

```diff
@@ -414,50 +414,43 @@ func WaitGroupChan(wg *sync.WaitGroup) <-chan struct{} {
 	return chAwait
 }
 
-// WithCloseChan wraps a context so that it is canceled if the passed in
-// channel is closed.
-// NOTE: Spins up a goroutine that exits on cancellation.
-// REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (ctx context.Context, cancel context.CancelFunc) {
-	ctx, cancel = context.WithCancel(parentCtx)
+// WithCloseChan wraps a context so that it is canceled if the passed in channel is closed.
+// Deprecated: Call StopChan.Ctx directly
+func WithCloseChan(parentCtx context.Context, chStop chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).Ctx(parentCtx)
+}
 
-	go func() {
-		select {
-		case <-chStop:
-		case <-ctx.Done():
-		}
-		cancel()
-	}()
+// ContextFromChan creates a context that finishes when the provided channel receives or is closed.
+// Deprecated: Call StopChan.NewCtx directly.
+func ContextFromChan(chStop chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).NewCtx()
+}
 
-	return ctx, cancel
+// ContextFromChanWithTimeout creates a context with a timeout that finishes when the provided channel receives or is closed.
+// Deprecated: Call StopChan.CtxCancel directly
+func ContextFromChanWithTimeout(chStop chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CtxCancel(context.WithTimeout(context.Background(), timeout))
 }
 
-// ContextFromChan creates a context that finishes when the provided channel
-// receives or is closed.
-// When channel closes, the ctx.Err() will always be context.Canceled
-// NOTE: Spins up a goroutine that exits on cancellation.
-// REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		select {
-		case <-chStop:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	return ctx, cancel
+// A StopChan signals when some work should stop.
+type StopChan chan struct{}
```

Comment:
this is much clearer now without the constructor and the stop func.

looks good and makes sense to me now